src/ray/raylet/node_manager.cc

```diff
@@ -210,14 +210,46 @@ ray::Status NodeManager::RegisterGcs() {
   return ray::Status::OK();
 }
 
+void NodeManager::KillWorker(std::shared_ptr<Worker> worker) {
+  /* If we're just cleaning up a single worker, allow it some time to clean
+    * up its state before force killing. The client socket will be closed
+    * and the worker struct will be freed after the timeout. */
+  kill(worker->Pid(), SIGTERM);
+
+  auto retry_timer = std::make_shared<boost::asio::deadline_timer>(io_service_);
+  auto retry_duration = boost::posix_time::milliseconds(
+      RayConfig::instance().kill_worker_timeout_milliseconds());
+  retry_timer->expires_from_now(retry_duration);
+  retry_timer->async_wait([worker](const boost::system::error_code &error) {
+    // Force kill driver.
+    kill(worker->Pid(), SIGKILL);
```

Comment:
what will happen if the worker is already dead?

Reply:
In that case, node manager has already received DisconnectClient message and corresponding cleanup for the worker has been done.

Reply:
Yes it's ok to send to a dead process. 
And if a worker dies before the driver, then the cleanup of this worker has already been handled when it receives DisconnectMessage for the worker, and the worker will be removed from worker_pool so it won't enter here.